cpp/src/Slice/Scanner.l

```diff
@@ -552,154 +548,150 @@ floating_literal    (({fractional_constant}{exponent_part}?)|((\+|-)?{dec}+{expo
-    keywordMap["short"] = ICE_SHORT;
-    keywordMap["int"] = ICE_INT;
-    keywordMap["long"] = ICE_LONG;
-    keywordMap["float"] = ICE_FLOAT;
-    keywordMap["double"] = ICE_DOUBLE;
-    keywordMap["string"] = ICE_STRING;
-    keywordMap["Object"] = ICE_OBJECT;
-    keywordMap["const"] = ICE_CONST;
-    keywordMap["false"] = ICE_FALSE;
-    keywordMap["true"] = ICE_TRUE;
-    keywordMap["idempotent"] = ICE_IDEMPOTENT;
-    keywordMap["tag"] = ICE_TAG;
-    // 'optional' is kept as an alias for 'tag' for backwards compatability.
-    // We need a separate token type since we infer 'optional T' to mean 'tag T?'.
-    // But for 'tag' we require an optional (nullable) type. No inferencing is done.
-    keywordMap["optional"] = ICE_OPTIONAL;
-    keywordMap["Value"] = ICE_VALUE;
-}
-
-// This function is always called directly after a match has been made, but directly before it's action block is run.
-void preAction()
-{
-    yycolno += yyleng;
+    void endLocation(TokenContext* location)
+    {
+        location->lastLine = yylineno;
+        location->lastColumn = yycolno;
+    }
 
-    // We only use the 'INITIAL' state to consume BOMs, which can only validly be the first match in a file. This
-    // function being called means a match has already been made, so we switch states since BOMs are no longer valid.
-    if(YY_START == INITIAL)
+    // This function is always called once, right before scanning begins.
+    void initScanner()
     {
-        BEGIN(PRE_SLICE);
+        // Ensure the scanner starts at line number 1, column position 0.
+        yylineno = 1;
+
+        keywordMap["module"] = ICE_MODULE;
+        keywordMap["class"] = ICE_CLASS;
+        keywordMap["interface"] = ICE_INTERFACE;
+        keywordMap["exception"] = ICE_EXCEPTION;
+        keywordMap["struct"] = ICE_STRUCT;
+        keywordMap["sequence"] = ICE_SEQUENCE;
+        keywordMap["dictionary"] = ICE_DICTIONARY;
+        keywordMap["enum"] = ICE_ENUM;
+        keywordMap["out"] = ICE_OUT;
+        keywordMap["extends"] = ICE_EXTENDS;
+        keywordMap["throws"] = ICE_THROWS;
+        keywordMap["void"] = ICE_VOID;
+        keywordMap["byte"] = ICE_BYTE;
+        keywordMap["bool"] = ICE_BOOL;
+        keywordMap["short"] = ICE_SHORT;
+        keywordMap["int"] = ICE_INT;
+        keywordMap["long"] = ICE_LONG;
+        keywordMap["float"] = ICE_FLOAT;
+        keywordMap["double"] = ICE_DOUBLE;
+        keywordMap["string"] = ICE_STRING;
+        keywordMap["Object"] = ICE_OBJECT;
+        keywordMap["const"] = ICE_CONST;
+        keywordMap["false"] = ICE_FALSE;
+        keywordMap["true"] = ICE_TRUE;
+        keywordMap["idempotent"] = ICE_IDEMPOTENT;
+        keywordMap["tag"] = ICE_TAG;
+        // 'optional' is kept as an alias for 'tag' for backwards compatability.
+        // We need a separate token type since we infer 'optional T' to mean 'tag T?'.
+        // But for 'tag' we require an optional (nullable) type. No inferencing is done.
```

Comment:
Not related to the PR, but we need to rework this comment, I think it comes from the 4.0 merge. We don't have nullable types in Ice 3.8.

Reply:
Yes, I just went ahead and fixed it (by removing the last two sentences).
Might as well do it now.